src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -480,6 +461,108 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
             _incomingBuffer.Discard(frameHeader.Length);
         }
 
+        private async void SendSettingsAck()
```

Comment:
Just out of curiosity, what is the advantage to having these be `async void`?

Reply:
In general `async void` should never be used.  The only exception is for UI button click event handlers, etc.  It should be `async Task`.

cc: @stephentoub

Reply:
The reason for `async void` here and elsewhere is that the callers don't want to wait for completion of the async operation. In this specific case, we don't want to stop processing the incoming frames just because someone else has the _writerLock (or whatever causes us to go async). We just want to ensure that the SETTINGS ack (or RST_STREAM etc) will be sent eventually; if we can do it right away, great, and if not, it will run eventually.

Alternatively, I could 
(a) return Task here and ignore it. This seems like a bad coding pattern to me -- the returned Task should basically never be used.
(b) return Task, and then use Task.Run(SendSettingsAck) to queue these to the threadpool. This seems like unnecessary overhead in the common case where these things complete synchronously.
(c) something else?

I'll defer to whatever pattern @stephentoub recommends here.

Reply:
https://stackoverflow.com/questions/45447955/why-exactly-is-void-async-bad

https://theburningmonk.com/2012/10/c-beware-of-async-void-in-your-code/

Reply:
Yes, I'm aware of that. What do you propose instead? Return Task and ignore it? It seems rather awkward to return a Task that absolutely must not be awaited.

(I should be ensuring that exceptions don't leak out here and kill the process; I'll defer making any changes here until we decide on the general approach.)

Reply:
There seems to be a lot written about the dangers of async void, but not a lot of info about when it is actually useful. I think we should see what Stephen says, but if we don't hear from him I'm in favor of option (b) just because our ability to handle errors in these methods is so restricted.

Reply:
I have three main issues with "async void":
1. You're initiating an asynchronous operation and not providing back any handle that can be used to join with the work.  (And if your method isn't named with an "Async" suffix, then there's no indication that it's doing so.)
2. You're interacting with the current SynchronizationContext, such that calling the method does the equivalent of `_capturedContext = SynchronizationContext.Current; _capturedContext?.OperationStarted();` and finishing the method does the equivalent of `_capturedContext?.OperationCompleted();`.
3. Related to the second, any unhandled exception inside of the async method is going to be propagated back to that captured context, whatever it may have been.

(1) is inherent to anything that's fire-and-forget, as it's mainly a complaint about fire-and-forget rather than about the specific mechanism used to achieve that.  I wouldn't recommend queueing to the thread pool just to work around that, since, beyond the extra overhead, that doesn't change its behavior as fire-and-forget.  Queueing to the thread pool does, however, impact (2) and (3), as there won't be a context to capture on the thread pool (it'll always be null or the default instance).

I would recommend just having the methods return `Task`, and then be explicit about ignoring the Task at the call site:
```C#
Task ignored = SendSettingsAckAsync();
```
(also with an `Async` suffix).  That way, it's the point of the fire-and-forget that's explicit about doing so (it's clear in a code review), we don't have unnecessary interaction with synchronization context, etc.  The main downside with this is that if there's an exception that goes unhandled inside SendSettingsAckAsync, it's going to get eaten, so I'd suggest SendSettingsAckAsync itself never allow that to happen, e.g. it should have a try/catch block around its whole implementation, and the catch block should do whatever's appropriate with an exception, e.g. logging it, fail fasting, throwing it on the thread pool to crash, etc.

Reply:
Okay, I can do that.

I think I also need to modify the error handling here in general. A failed write to the underlying stream needs to abort the connection and fail any subsequent write attempts. I'll update the logic to include that.